### PR TITLE
in case curl is not installed or installed,  but does not support https, use wget to download envoy

### PR DESCRIPTION
**Pull request by vadimeisenbergibm**
check if curl supports https and check if wget is installed
if neither of the above is true, print an error message and exit

Sometimes curl can be installed without https support.

### CODE REVIEW

File: bin/init.sh [hunk 1]
```diff
@@ -56,7 +56,18 @@ if [ ! -f vendor/envoy-$PROXYVERSION ] ; then
...
     pushd $OUT
     # New version of envoy downloaded. Save it to cache, and clean any old version.
-    curl -Lo - https://storage.googleapis.com/istio-build/proxy/envoy-$PROXY.tar.gz | tar xz
+
+    if curl --version | grep Protocols  | grep https; then
+	DOWNLOAD_COMMAND='curl -Lo -'
+    elif command -v wget > /dev/null; then
+	DOWNLOAD_COMMAND='wget -qO -'
+    else
+	echo Error: curl does not support https and wget is not installed. \
```
Comment (rshriram) [inline bin/init.sh]: ? curl always supported https. You can use curl -k to skip cert verification

**Comment by ldemailly** [comment]
> From what I see on Ubuntu 16.04.3 LTS, sudo apt-get install curl installs it without https support.

that doesn't sound right

do you have a specific error you saw and can share ?

```
Laurents-MBP:go-ui dl$ docker run -ti ubuntu:xenial
root@4e4b88021dc2:/# curl
bash: curl: command not found
root@4e4b88021dc2:/# apt-get update && apt-get install curl
[...]
The following additional packages will be installed:
  ca-certificates krb5-locales libasn1-8-heimdal libcurl3-gnutls libffi6 libgmp10 libgnutls30 libgssapi-krb5-2 libgssapi3-heimdal libhcrypto4-heimdal
  libheimbase1-heimdal libheimntlm0-heimdal libhogweed4 libhx509-5-heimdal libidn11 libk5crypto3 libkeyutils1 libkrb5-26-heimdal libkrb5-3
  libkrb5support0 libldap-2.4-2 libnettle6 libp11-kit0 libroken18-heimdal librtmp1 libsasl2-2 libsasl2-modules libsasl2-modules-db libsqlite3-0
  libssl1.0.0 libtasn1-6 libwind0-heimdal openssl
Suggested packages:
  gnutls-bin krb5-doc krb5-user libsasl2-modules-otp libsasl2-modules-ldap libsasl2-modules-sql libsasl2-modules-gssapi-mit
  | libsasl2-modules-gssapi-heimdal
The following NEW packages will be installed:
  ca-certificates curl krb5-locales libasn1-8-heimdal libcurl3-gnutls libffi6 libgmp10 libgnutls30 libgssapi-krb5-2 libgssapi3-heimdal
  libhcrypto4-heimdal libheimbase1-heimdal libheimntlm0-heimdal libhogweed4 libhx509-5-heimdal libidn11 libk5crypto3 libkeyutils1 libkrb5-26-heimdal
  libkrb5-3 libkrb5support0 libldap-2.4-2 libnettle6 libp11-kit0 libroken18-heimdal librtmp1 libsasl2-2 libsasl2-modules libsasl2-modules-db
  libsqlite3-0 libssl1.0.0 libtasn1-6 libwind0-heimdal openssl
0 upgraded, 34 newly installed, 0 to remove and 19 not upgraded.
Need to get 5343 kB of archives.
After this operation, 19.0 MB of additional disk space will be used.
Do you want to continue? [Y/n] y
[...]
Processing triggers for ca-certificates (20170717~16.04.1) ...
Updating certificates in /etc/ssl/certs...
148 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
```
^^^

**Comment by vadimeisenbergibm** [comment]
@rshriram @ldemailly this is what I have on my ubuntu VM:
```
vadime@vadim-istio:~$ curl --version
curl 7.54.0 (x86_64-pc-linux-gnu) libcurl/7.54.0 zlib/1.2.8 nghttp2/1.7.1
Protocols: dict file ftp gopher http imap pop3 rtsp smtp telnet tftp 
Features: IPv6 Largefile libz HTTP2 UnixSockets 
vadime@vadim-istio:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial
vadime@vadim-istio:~$ curl -Lo - https://storage.googleapis.com/istio-build/proxy/envoy-$PROXY.tar.gz 
curl: (1) Protocol "https" not supported or disabled in libcurl
```

**Comment by vadimeisenbergibm** [comment]
Changed the description of the PR:
> Sometimes curl can be installed without https support.

**Comment by ldemailly** [comment]
can you run apt-get update && apt-get upgrade curl on your vm and paste the output ?

**Comment by vadimeisenbergibm** [comment]
@ldemailly 
```
sudo apt-get update && sudo apt-get upgrade curl 
[sudo] password for vadime: 
Hit:1 http://ppa.launchpad.net/webupd8team/atom/ubuntu xenial InRelease
Hit:2 http://il.archive.ubuntu.com/ubuntu xenial InRelease          
Get:3 http://il.archive.ubuntu.com/ubuntu xenial-updates InRelease [102 kB]                    
Hit:4 http://storage.googleapis.com/bazel-apt stable InRelease                                           
Get:5 http://security.ubuntu.com/ubuntu xenial-security InRelease [102 kB]                               
Get:6 http://il.archive.ubuntu.com/ubuntu xenial-backports InRelease [102 kB]      
Get:7 http://il.archive.ubuntu.com/ubuntu xenial-updates/main amd64 DEP-11 Metadata [307 kB] 
Get:8 http://il.archive.ubuntu.com/ubuntu xenial-updates/main DEP-11 64x64 Icons [222 kB]           
Get:9 http://il.archive.ubuntu.com/ubuntu xenial-updates/universe amd64 DEP-11 Metadata [185 kB]   
Hit:10 https://download.docker.com/linux/ubuntu xenial InRelease                                
Get:11 http://il.archive.ubuntu.com/ubuntu xenial-updates/universe DEP-11 64x64 Icons [253 kB]     
Get:12 http://il.archive.ubuntu.com/ubuntu xenial-updates/multiverse amd64 DEP-11 Metadata [5,892 B]                  
Get:13 http://il.archive.ubuntu.com/ubuntu xenial-backports/main amd64 DEP-11 Metadata [3,328 B]                          
Get:14 http://il.archive.ubuntu.com/ubuntu xenial-backports/universe amd64 DEP-11 Metadata [4,716 B]                      
Get:15 http://security.ubuntu.com/ubuntu xenial-security/main amd64 DEP-11 Metadata [62.8 kB]                                
Get:16 http://security.ubuntu.com/ubuntu xenial-security/main DEP-11 64x64 Icons [71.2 kB]
Get:17 http://security.ubuntu.com/ubuntu xenial-security/universe amd64 DEP-11 Metadata [51.4 kB]
Get:18 http://security.ubuntu.com/ubuntu xenial-security/universe DEP-11 64x64 Icons [85.1 kB]
Fetched 1,558 kB in 1s (965 kB/s)                                      
Reading package lists... Done
Reading package lists... Done
Building dependency tree       
Reading state information... Done
curl is already the newest version (7.47.0-1ubuntu2.5).
Calculating upgrade... Done
The following packages have been kept back:
  libdrm-amdgpu1 libdrm2 libegl1-mesa libgbm1 libgl1-mesa-dri libinput10 libwayland-egl1-mesa libxatracker2 xserver-xorg-core-hwe-16.04 xserver-xorg-hwe-16.04
  xserver-xorg-input-evdev-hwe-16.04 xserver-xorg-input-synaptics-hwe-16.04 xserver-xorg-input-wacom-hwe-16.04 xserver-xorg-video-amdgpu-hwe-16.04
  xserver-xorg-video-ati-hwe-16.04 xserver-xorg-video-fbdev-hwe-16.04 xserver-xorg-video-intel-hwe-16.04 xserver-xorg-video-nouveau-hwe-16.04
  xserver-xorg-video-qxl-hwe-16.04 xserver-xorg-video-radeon-hwe-16.04 xserver-xorg-video-vesa-hwe-16.04 xserver-xorg-video-vmware-hwe-16.04
The following packages will be upgraded:
  apparmor atom bazel ca-certificates-java cpp-5 cracklib-runtime distro-info-data dnsmasq-base docker-ce dpkg dpkg-dev g++-5 gcc-5 gcc-5-base
  gir1.2-packagekitglib-1.0 gnome-software gnome-software-common golang-1.6-go golang-1.6-src grub-common grub-pc grub-pc-bin grub2-common initramfs-tools
  initramfs-tools-bin initramfs-tools-core iproute2 ipxe-qemu klibc-utils libapparmor-perl libapparmor1 libasan2 libatomic1 libcc1-0 libcilkrts5 libcrack2
  libcryptsetup4 libdpkg-perl libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libgcc-5-dev libgl1-mesa-glx libglapi-mesa libgnutls-openssl27 libgnutls30 libgomp1
  libgweather-3-6 libgweather-common libitm1 libklibc liblsan0 libmpx0 libnm-glib-vpn1 libnm-glib4 libnm-gtk-common libnm-gtk0 libnm-util2 libnm0 libnma-common
  libnma0 libnux-4.0-0 libnux-4.0-common libpackagekit-glib2-16 libpam-systemd libplymouth4 libpulse-mainloop-glib0 libpulse0 libpulsedsp libpython2.7
  libpython2.7-dev libpython2.7-minimal libpython2.7-stdlib libquadmath0 librados2 librbd1 libseccomp2 libsnapd-glib1 libstdc++-5-dev libstdc++6 libsystemd0
  libtsan0 libubsan0 libudev1 libunity-control-center1 light-themes lshw network-manager network-manager-gnome nux-tools plymouth plymouth-label
  plymouth-theme-ubuntu-logo plymouth-theme-ubuntu-text pulseaudio pulseaudio-module-bluetooth pulseaudio-module-x11 pulseaudio-utils python2.7 python2.7-dev
  python2.7-minimal python3-distupgrade python3-update-manager resolvconf snapd snapd-login-service squashfs-tools suru-icon-theme systemd systemd-sysv
  ubuntu-artwork ubuntu-core-launcher ubuntu-desktop ubuntu-minimal ubuntu-mobile-icons ubuntu-mono ubuntu-release-upgrader-core ubuntu-release-upgrader-gtk
  ubuntu-software ubuntu-standard udev unattended-upgrades unity-control-center unity-control-center-faces update-manager update-manager-core
126 upgraded, 0 newly installed, 0 to remove and 22 not upgraded.
Need to get 346 MB of archives.
After this operation, 53.6 MB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Get:1 http://il.archive.ubuntu.com/ubuntu xenial-updates/main amd64 dpkg amd64 1.18.4ubuntu1.3 [2,084 kB]
...
Processing triggers for libc-bin (2.23-0ubuntu10) ...
Processing triggers for resolvconf (1.78ubuntu5) ...
Processing triggers for ca-certificates (20170717~16.04.1) ...
Updating certificates in /etc/ssl/certs...
0 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...

done.
done.
vadime@vadim-istio:~/go/src/istio.io/istio$ curl --version
curl 7.54.0 (x86_64-pc-linux-gnu) libcurl/7.54.0 zlib/1.2.8 nghttp2/1.7.1
Protocols: dict file ftp gopher http imap pop3 rtsp smtp telnet tftp 
Features: IPv6 Largefile libz HTTP2 UnixSockets 
vadime@vadim-istio:~/go/src/istio.io/istio$ curl -Lo - https://storage.googleapis.com/istio-build/proxy/envoy-$PROXY.tar.gz
curl: (1) Protocol "https" not supported or disabled in libcurl
```

**Comment by vadimeisenbergibm** [comment]
@ldemailly @rshriram Additional Info - the VM was installed from an Ubuntu ISO. I do not remember if I installed curl explicitly or it was preinstalled with the ISO. But the point is this - there are installations of Unix where curl does not support HTTPS.

Googling "curl: (1) Protocol "https" not supported or disabled in libcurl" returns 6400 results. So it is not only my VM, it could happen to other people too. If there is a possibility of curl not supporting HTTPS, why shouldn't we check it in our script build, instead of just failing in these cases?

**Comment by vadimeisenbergibm** [comment]
Added explicit check that curl is installed, and also explicit messages.

**Comment by ldemailly** [comment]
mine says
```
# curl --version
curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets 
```

maybe it's some export thing. surprising it would work for wget though, in a way I'd prefer to get to the bottom of it than switch downloader (or maybe we could use a go program)

**Comment by vadimeisenbergibm** [comment]
Additional option is just to exit and to ask the user to install get and to add https support, without falling back to wget.

File: bin/init.sh [hunk 2]
```diff
@@ -56,7 +56,33 @@ if [ ! -f vendor/envoy-$PROXYVERSION ] ; then
     mkdir -p $OUT
     pushd $OUT
     # New version of envoy downloaded. Save it to cache, and clean any old version.
-    curl -Lo - https://storage.googleapis.com/istio-build/proxy/envoy-$PROXY.tar.gz | tar xz
+
+    DOWNLOAD_COMMAND=""
+    if command -v curl > /dev/null; then
```
Comment (ldemailly) [inline bin/init.sh]: I've never heard of 'command' before, I think it would work the same without it ?

Comment (vadimeisenbergibm) [reply bin/init.sh]: Here it is http://pubs.opengroup.org/onlinepubs/9699919799/utilities/command.html, POSIX standard.

I found it in this stackoverflow answer https://stackoverflow.com/a/677212/6028420.

**Pull request merged**